pennylane/ops/functions/bind_new_parameters.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the qml.bind_new_parameters function.
+"""
+
+from typing import Sequence
+import copy
+from functools import singledispatch
+
+import pennylane as qml
+from pennylane.typing import TensorLike
+from pennylane.operation import Operator, Tensor
+
+from ..op_math import CompositeOp, SymbolicOp, ScalarSymbolicOp, Adjoint, Pow, SProd
+
+
+@singledispatch
+def bind_new_parameters(op: Operator, params: Sequence[TensorLike]) -> Operator:
+    """Create a new operator with updated parameters
+
+    This function takes an :class:`~.Operator` and new parameters as input and
+    returns a new operator of the same type with the new parameters. This function
+    does not mutate the original operator.
+
+    Args:
+        op (.Operator): Operator to update
+        params (Sequence[TensorLike]): New parameters to create operator with
+
+    Returns:
+        .Operator: New operator with updated parameters
+
+    Raises:
+        ValueError: If the shape of the old and new operator parameters don't match
+    """
+
```

Comment:
We do not need to decide on an answer to this question in this PR, but we should think about the behaviour for later.

What do we want this function to do for an operator with no parameters?

bind_new_parameters(PauliX(0), tuple())
Do we need to create a new operator in this case, or can we simply return the original operation?

Reply:
is this function just meant to be for user convenience? it seems weird to call `bind_new_parameters` on a non-parametric op, but if we intend on using it in a fairly automated way then I guess it might happen. I think making a copy is fine for now, but we can add `if op.num_params == 0: return op` in the default constructor if we anticipate it more

Reply:
This will be a developer function.

In testing its use on a large `Sum`:
```
data = qml.data.load("qchem", molname="LiH", basis="STO-3G", bondlength=1.5)[0]
H = data.hamiltonian
s = qml.pauli.pauli_sentence(H).operation()

new_params = [0.5 ] * 631
bind_new_parameters(s, new_params)
```

I discovered an edge case for `Identity`.  

The problem with the `Identity` operation can be solved by returning the original object if there's no new parameters.